### KAD analysis for two versions of *E. coli* genome assemblies
#### KAD profiling
**Running script - KAD profiling**
```
perl KADprofile.pl \
     --read MG1655_1.fq.gz --read MG1655_2.fq.gz \
     --asm U00096.1.fasta --aid U00096.1 \
     --asm U00096.3.fasta --aid U00096.3  \
     --prefix MG1655
```
**Outputs in the output directory (MG1655)**
├── figures  
│   ├── reads.kmer.abundance.profile.pdf  
│   ├── U00096.1.kad.profile.pdf  
│   └── U00096.3.kad.profile.pdf  
├── MG1655_0_reads.kmer.table.txt  
├── MG1655_1_U00096.1.kmer.table.txt  
├── MG1655_1_U00096.3.kmer.table.txt  
├── MG1655_2_merge.kmer.table.txt  
├── MG1655_3_readkmer.cmode  
├── MG1655_4_kad.txt  
├── MG1655_5_readkmer.counts.txt  
├── MG1655_6_bincount.txt  
├── MG1655_7_kad.stat.txt  
├── MG1655_X_Rmd.render.R  
└── report  
    └── MG1655_KADprofile.report.html

The major outputs include
1. KAD result in MG1655_4_kad.txt and here are columns:  
Kmer: k-mer sequence  
reads: abundance of the k-mer in reads  
U00096.1: copy number of the k-mer in the assembly of U00096.1  
U00096.3: copy number of the k-mer in the assembly of U00096.3  
U00096.1.KAD: the KAD value of the k-mer of U00096.1  
U00096.3.KAD: the KAD value of the k-mer of U00096.3  

2. KAD summary in MG1655_7_kad.stat.txt.  
3. A report in HTML format "MG1655_KADprofile.report.html" in the directory of "report"  
4. Editable PDF figures in the directory of "figures"

Here are the profiling plot for the assembly U00096.1:  
![KAD profiling of U00096.1](https://github.com/liu3zhenlab/KAD/raw/master/cases/Ecoli/analysis/MG1655/figures/U00096.1.kad.profile.png)

#### KAD distribution on contigs or chromosomes  
**Running script - KAD landscape distribution**
```
perl KADdist.pl \
     --prefix U00096.1dist \
     --kad ./MG1655/MG1655_4_kad.txt \
     --aid U00096.1 --asm U00096.1.fasta
```
**Outputs in the output directory (U00096.1dist)**
├── pdf  
│   └── U00096.1.kad.dist.pdf  
├── U00096.1dist_1_kmer.fasta  
├── U00096.1dist_2_kmer.aln  
├── U00096.1dist_3_kmer.kad.bed  
├── U00096.1dist_4_error.pos.bed  
├── U00096.1dist_4_error.pos.merge.bed  
├── U00096.1dist_5_asm.lengths  
├── U00096.1dist_5_kad.bigwig  
├── U00096.1dist_5_kad.wig  
├── U00096.1dist_6_KADtype.dist.txt  
└── U00096.1dist_KADdist.log

The file "U00096.1dist_6_KADtype.dist.txt" contains number of k-mers for each k-mer group (Good, Error, OverRep, LowUnderRep (lounrep), and HighUnderRep (hiunrep), and uncategorized (others)). The data in this file are used to produce the KAD landscape distrition plot "U00096.1.kad.dist.pdf". Exact positions of error k-mers can be found in two BED files "U00096.1dist_4_error.pos.bed" and "U00096.1dist_4_error.pos.merge.bed". The former BED file includes all error k-mers. The latter one has merged data if neighbor error k-mers overlap, which are referred to as error regions.

Here are the landscape plot for the assembly U00096.1:  
![KAD landscape plot](https://github.com/liu3zhenlab/KAD/blob/master/cases/Ecoli/analysis/U00096.1dist/U00096.1.kad.dist.png)  

**The landscape distribution can be generated for U00096.3 using the following script**
```
perl KADdist.pl \
     --prefix U00096.1dist \
     --kad ./MG1655/MG1655_4_kad.txt \
     --aid U00096.3 --asm U00096.3.fasta
```

#### KAD comparison between two assemblies
**KAD comparison of two assemblies using the output from KAD profiling**
```
perl KADcompare.pl \
     --set1 U00096.1
     --set2 U00096.3
     --prefix U00096.1vs3
     MG1655/MG1655_4_kad.txt
```

**Outputs in the output directory (U00096.1vs3)**
.
├── report  
│   └── U00096.1vs3_U00096.1-U00096.3.report.html  
├── U00096.1vs3_1_U00096.1-U00096.3.KADdiff.txt  
├── U00096.1vs3_2_U00096.1-U00096.3.bincount.txt  
├── U00096.1vs3_2_U00096.1-U00096.3.Fig1.raw.pdf  
├── U00096.1vs3_2_U00096.1-U00096.3.Fig2.cuberoot.pdf  
└── U00096.1vs3_X_Rmd.render.R

In the output, "U00096.1vs3_1_U00096.1-U00096.3.KADdiff.txt" lists all k-mers with distinct KAD values of the two assemblies. "U00096.1vs3_2_U00096.1-U00096.3.bincount.txt" include counts of k-mers with distinct KADs in each small interval (bincount) for each of the assembly. Figures are visualized version of the bincount data in the raw and cuberoot scales.

Here are the KAD comparison plot for the two assemblies U00096.1 and U00096.2:  
![KAD comparison plot](https://github.com/liu3zhenlab/KAD/blob/master/cases/Ecoli/analysis/U00096.1vs3/U00096.1vs3_2_U00096.1-U00096.3.Fig1.raw.png =200x200)  

